flink/src/main/java/org/apache/iceberg/flink/actions/MigrateAction.java

```diff
@@ -0,0 +1,374 @@
+    try {
+      hiveTable = flinkHiveCatalog.getHiveTable(tableSource);
+    } catch (TableNotExistException e) {
+      throw new RuntimeException(String.format("The source table %s not exists ! ", hiveSourceTableName));
+    }
+
+    List<FieldSchema> fieldSchemaList = hiveTable.getSd().getCols();
+    List<FieldSchema> partitionList = hiveTable.getPartitionKeys();
+    fieldSchemaList.addAll(partitionList);
+    Schema icebergSchema = HiveSchemaUtil.convert(fieldSchemaList);
+    PartitionSpec spec = toPartitionSpec(partitionList, icebergSchema);
+
+    String hiveFormat = hiveTable.getSd().getInputFormat();
+    FileFormat fileFormat;
+    switch (hiveFormat) {
+      case PATQUET_INPUT_FORMAT:
+        fileFormat = FileFormat.PARQUET;
+        break;
+
+      case ORC_INPUT_FORMAT:
+        fileFormat = FileFormat.ORC;
+        break;
+
+      case AVRO_INPUT_FORMAT:
+        fileFormat = FileFormat.AVRO;
+        break;
+
+      default:
+        throw new UnsupportedOperationException("Unsupported file format");
+    }
+
+    TableIdentifier identifier = TableIdentifier.of(toNamespace(baseNamespace), icebergTableName);
+    String hiveLocation = hiveTable.getSd().getLocation();
+
+    Map<String, String> properties = Maps.newHashMap();
+    properties.put(TableProperties.DEFAULT_FILE_FORMAT, fileFormat.name());
+    if (!baseNamespace.isEmpty()) {
+      properties.put(FlinkCatalogFactory.BASE_NAMESPACE, baseNamespace.toString());
+    }
+
+    Table icebergTable;
+    if (icebergCatalog instanceof HadoopCatalog) {
+      icebergTable = icebergCatalog.createTable(identifier, icebergSchema, spec, properties);
+    } else {
+      icebergTable = icebergCatalog.createTable(identifier, icebergSchema, spec, hiveLocation, properties);
+    }
+
+    String nameMapping =
+        PropertyUtil.propertyAsString(icebergTable.properties(), TableProperties.DEFAULT_NAME_MAPPING, null);
+    MetricsConfig metricsConfig = MetricsConfig.fromProperties(icebergTable.properties());
+
+    List<DataFile> files = null;
+    if (spec.isUnpartitioned()) {
+      MigrateMap migrateMap = new MigrateMap(spec, nameMapping, fileFormat, metricsConfig);
+      DataStream<PartitionAndLocation> dataStream =
+          env.fromElements(new PartitionAndLocation(hiveLocation, Maps.newHashMap()));
+      DataStream<List<DataFile>> ds = dataStream.map(migrateMap);
+      try {
+        files = Lists.newArrayList(ds.executeAndCollect("migrate table :" + icebergTable.name())).stream()
+            .flatMap(Collection::stream).collect(Collectors.toList());
+      } catch (Exception e) {
+        throw new RuntimeException(e);
+      }
+    } else {
+      try {
+        files = importPartitions(spec, tableSource, nameMapping, fileFormat, metricsConfig, icebergTable.name());
+      } catch (Exception e) {
+        throw new RuntimeException(e);
+      }
+    }
+
+    AppendFiles append = icebergTable.newAppend();
+    files.forEach(append::appendFile);
```

Comment:
I am concerned about how this works for huge tables with millions of files.

- collecting all of them into a single list can cause memory pressure.
- writing all of them into a single manifest file may also hurt read performance for partitioned tables, as we can't filter manifest files by partition value ranges in manifest list file. 

Maybe we should commit the files by batches. Each Hive partition is a batch and all data files under one Hive partitions get written into a single manifest file. We can use `AppendFiles#appendManifest` API too.

Reply:
thanks for  your suggesion，I will refactor this logic

Reply:
> * collecting all of them into a single list can cause memory pressure.

I looked at the code again, the `listStatus` method of hadoop filesystem use `ArrayList` to construct the `FileStatus` array.
```
  public FileStatus[] listStatus(Path f, PathFilter filter) 
                                   throws FileNotFoundException, IOException {
    ArrayList<FileStatus> results = new ArrayList<FileStatus>();
    listStatus(results, f, filter);
    return results.toArray(new FileStatus[results.size()]);
  }
```

although there is the `org.apache.hadoop.fs.FileSystem#listStatusIterator` method, the `listStatus` method is still called in the internal of `FileSystem#listStatusIterator` method, so when we get the file list of hive, even if we use Iterator, I think it is still useless.

Reply:
After the refactoring of emitting ManifestFile per partition, we are only listing the files under one partition. Then I won't be concerned for the list. Previously, we are collecting all files under a table.